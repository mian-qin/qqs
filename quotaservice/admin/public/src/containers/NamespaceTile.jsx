import PropTypes from 'prop-types';
import React from 'react';
import { Component } from 'react';

export default class NamespaceTile extends Component {
  constructor() {
    super();
    this.state = {};
  }

  canMakeChanges() {
    // permissive by default, to maintain previous functionality in case of misconfiguration
    return this.state.canMakeChanges !== false;
  }

  onClick = () => {
    const { namespace, selectNamespace, capabilitiesEnabled } = this.props;
    selectNamespace(namespace.name, capabilitiesEnabled ? this.canMakeChanges() : true);
  }

  fetchCanMakeChanges() {
    if ('canMakeChangesFetched' in this.state)
      return;

    const { namespace: { name: namespaceName }, eventEmitter } = this.props;
    const callback = (canMakeChanges = true) =>
      this.setState(prevState => ({ ...prevState, canMakeChanges, canMakeChangesFetched: true }));

    eventEmitter.dispatchEvent(new CustomEvent(
      'QuotaService.getCapabilities',
      { detail: { namespaceName, callback } }
    ));
  }

  componentWillMount() {
    const { capabilities, capabilitiesEnabled } = this.props;
    if (capabilitiesEnabled && !capabilities.inRequest && !capabilities.error) {
      this.fetchCanMakeChanges();
    }
  }

  render() {
    const { isSelected, namespace } = this.props;
    const buckets = namespace.buckets || {};
    const isGrayedOutBecauseCantMakeChanges = this.props.capabilitiesEnabled ? !this.canMakeChanges() : false;
    const className = 'flex-box flex-tile namespace'
      + (isSelected ? ' selected' : '')
      + (isGrayedOutBecauseCantMakeChanges ? ' disabled' : '');

    return (
      <div className={className} onClick={this.onClick}>
        <p className="title">{namespace.name}</p>
        <hr />
        {this.renderBucket(namespace.dynamic_bucket_template)}
        {this.renderBucket(namespace.default_bucket)}
        {Object.keys(buckets).map(key =>
          this.renderBucket(buckets[key])
        )}
      </div>
    )
  }

  renderBucket(bucket) {
    if (!bucket)
      return;

    return <div key={bucket.name} className="bucket">{bucket.name}</div>;
  }
}

NamespaceTile.propTypes = {
  isSelected: PropTypes.bool.isRequired,
  namespace: PropTypes.object.isRequired,
  capabilities: PropTypes.object.isRequired,
  eventEmitter: PropTypes.object.isRequired,
  capabilitiesEnabled: PropTypes.bool.isRequired,
  selectNamespace: PropTypes.func.isRequired
}
