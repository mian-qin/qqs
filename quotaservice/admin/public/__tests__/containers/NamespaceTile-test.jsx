import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';

import NamespaceTile from '../../src/containers/NamespaceTile.jsx';

describe('NamespaceTile', () => {
  let props, component;

  beforeEach(() => {
    props = {
      namespace: {
        name: 'namespace-name',
        dynamic_bucket_template: { name: 'dynamic_bucket_template-name' },
        default_bucket: { name: 'dynamic_bucket_template-name' },
        buckets: {
          bucket1: { name: 'bucket1-name' },
          bucket2: { name: 'bucket2-name' },
        }
      },
      selectNamespace: () => null,
      isSelected: false,
      capabilitiesEnabled: false,
      capabilities: {},
      eventEmitter: {
        dispatchEvent: jest.fn(),
      }
    };
  });

  it('renders with minimum data', () => {
    component = shallow(<NamespaceTile {...props} />);
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('renders in selected state', () => {
    props.isSelected = true;
    component = shallow(<NamespaceTile {...props} />);
    expect(toJSON(component)).toMatchSnapshot();
    expect(component.find('.namespace.selected')).toHaveLength(1);
  });

  describe('capabilities', () => {
    beforeEach(() => {
      props.capabilitiesEnabled = true;
      component = shallow(<NamespaceTile {...props} />);
    });

    it('renders ok when can not make changes', () => {
      component.setState({ canMakeChanges: false });
      expect(toJSON(component)).toMatchSnapshot();
      expect(props.eventEmitter.dispatchEvent.mock.calls).toHaveLength(1);
      expect(component.find('.namespace.disabled')).toHaveLength(1);
    });

    it('renders ok when able to make changes', () => {
      component.setState({ canMakeChanges: true });
      expect(toJSON(component)).toMatchSnapshot();
      expect(props.eventEmitter.dispatchEvent.mock.calls).toHaveLength(1);
      expect(component.find('.namespace.disabled')).toHaveLength(0);
    });
  });
});
