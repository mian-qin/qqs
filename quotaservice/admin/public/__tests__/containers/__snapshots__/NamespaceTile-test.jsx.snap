// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceTile capabilities renders ok when able to make changes 1`] = `
<div
  className="flex-box flex-tile namespace"
  onClick={[Function]}
>
  <p
    className="title"
  >
    namespace-name
  </p>
  <hr />
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="bucket1-name"
  >
    bucket1-name
  </div>
  <div
    className="bucket"
    key="bucket2-name"
  >
    bucket2-name
  </div>
</div>
`;

exports[`NamespaceTile capabilities renders ok when can not make changes 1`] = `
<div
  className="flex-box flex-tile namespace disabled"
  onClick={[Function]}
>
  <p
    className="title"
  >
    namespace-name
  </p>
  <hr />
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="bucket1-name"
  >
    bucket1-name
  </div>
  <div
    className="bucket"
    key="bucket2-name"
  >
    bucket2-name
  </div>
</div>
`;

exports[`NamespaceTile renders in selected state 1`] = `
<div
  className="flex-box flex-tile namespace selected"
  onClick={[Function]}
>
  <p
    className="title"
  >
    namespace-name
  </p>
  <hr />
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="bucket1-name"
  >
    bucket1-name
  </div>
  <div
    className="bucket"
    key="bucket2-name"
  >
    bucket2-name
  </div>
</div>
`;

exports[`NamespaceTile renders with minimum data 1`] = `
<div
  className="flex-box flex-tile namespace"
  onClick={[Function]}
>
  <p
    className="title"
  >
    namespace-name
  </p>
  <hr />
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="dynamic_bucket_template-name"
  >
    dynamic_bucket_template-name
  </div>
  <div
    className="bucket"
    key="bucket1-name"
  >
    bucket1-name
  </div>
  <div
    className="bucket"
    key="bucket2-name"
  >
    bucket2-name
  </div>
</div>
`;
